src/lib/client.js

```diff
@@ -26,18 +30,93 @@ module.exports = class BtpClient {
     this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
   }
 
+  tryToOpenWebSocketClient () {
+    return new Promise((resolve, reject) => {
+      const ws = new WebSocket(this._wsUri, {
+        perMessageDeflate: false
+      })
+      console.log('try to open', this._wsUri)
+      let hasBeenOpen = false
+      ws.on('open', () => {
+        console.log('open!')
+        hasBeenOpen = true
+        resolve(ws)
+      })
+      ws.on('error', (err) => {
+        if (hasBeenOpen) {
+          console.error('WebSocket error', err)
+        } else {
+          reject(err)
+        }
+      })
+      ws.on('close', () => {
+        console.log('close!')
+        // automatically reconnect if server reboots
+        if (hasBeenOpen && !ws.shouldClose) {
+          this.ensureUpstream()
+        }
+      })
+    })
+  }
+
+  getOpenWebSocketClient () {
+    return new Promise((resolve) => {
+      let done = false
+      const tryOnce = () => {
+        this.tryToOpenWebSocketClient().then(ws => {
+          if (done) { // this can happen if opening the WebSocket works, but just takes long
+            console.log('got it! but was already done')
+            ws.shouldClose = true
+            ws.close()
+          } else {
+            console.log('got it! setting done')
+            done = true
+            clearInterval(timer)
+            resolve(ws)
+          }
+        }).catch((err) => {
+          console.error('error caught while trying to open WebSocket client', err)
+        })
+      }
+      let timer = setInterval(tryOnce, CONNECT_RETRY_INTERVAL)
+      tryOnce()
+    })
+  }
+
+  ensureUpstream () {
+    return this.getOpenWebSocketClient().then(ws => {
+      console.log('got the client!')
+      this.ws = ws
+      ws.on('message', (msg) => {
+        this.msgHandler(msg)
+      })
+      this.incarnation++
+    }, (err) => {}) // eslint-disable-line handle-callback-err
+  }
+
   async connect () {
     debug('connecting to', this._wsUri)
-    const ws = new WebSocket(this._wsUri)
+    await this.getOpenWebSocketClient()
```

Comment:
`this.getOpenWebsocketClient` won't set `this.ws`. should this be `this.ensureUpstream`?

Reply:
ah, yes, good catch